### PR TITLE
NUnit_Engine: running methods extended and bugfixed

### LINKED ISSUE #435: NUnit TestSuite does not deserialise correctly
#### Description:

It crashes when `Settings` or `Properties` are not empty.

### LINKED ISSUE #436: Add methods to find NUnit test classes and methods etc.
#### Description:

As in the title, basic feature requested to enable browsing the tests programmatically.

### PR DESCRIPTION
**Pull request by pawelbaran**
### Issues addressed by this PR

 Closes #435
 Closes #436

 

 ### Test files

To be added.

 ### Changelog

 ### Additional comments

Not ideal to mix bug + feature, but this is a result of a wider exploration. Happy to split into separate PRs before merging along with targeted test scripts being added.

### CODE REVIEW

File: NUnit_Engine/Query/NunitTestClasses.cs [hunk 1]
```diff
@@ -20,20 +20,22 @@
...
-    [XmlRoot(ElementName = "assertions")]
-    public class Assertions : IObject
+    public static partial class Query
     {
-        [XmlElement(ElementName = "assertion")]
-        public virtual Assertion Assertion { get; set; }
+        [System.ComponentModel.Description("Returns all types in an assembly that contain any methods with NUnit 'Test' attribute.")]
+        [Input("assembly", "Assembly to parse in search for NUnit test types.")]
+        [Output("classes", "All types that contain methods with NUnit 'Test' attribute.")]
+        public static List<Type> NunitTestClasses(this Assembly assembly)
```
Comment (FraserGreenroyd) [inline NUnit_Engine/Query/NunitTestClasses.cs:36]:
```suggestion
        public static List<Type> NUnitTestClasses(this Assembly assembly)
```

File: NUnit_Engine/Query/NunitTestMethods.cs [hunk 2]
```diff
@@ -20,20 +20,31 @@
...
-    [XmlRoot(ElementName = "settings")]
-    public class Settings : IObject
+    public static partial class Query
     {
-        [XmlElement(ElementName = "setting")]
-        public virtual Setting Setting { get; set; }
+        [System.ComponentModel.Description("Returns all methods in an assembly that have NUnit 'Test' attribute.")]
+        [Input("assembly", "Assembly to parse in search for NUnit test methods.")]
+        [Output("methods", "All methods that have NUnit 'Test' attribute.")]
+        public static List<MethodInfo> NunitTestMethods(this Assembly assembly)
```
Comment (FraserGreenroyd) [inline NUnit_Engine/Query/NunitTestMethods.cs:37]:
```suggestion
        public static List<MethodInfo> NUnitTestMethods(this Assembly assembly)
```

File: NUnit_Engine/Query/NunitTestMethods.cs [hunk 3]
```diff
@@ -20,20 +20,31 @@
...
+        {
+            return assembly.GetTypes().SelectMany(x => x.NunitTestMethods()).ToList();
```
Comment (FraserGreenroyd) [inline NUnit_Engine/Query/NunitTestMethods.cs:39]:
```suggestion
            return assembly.GetTypes().SelectMany(x => x.NUnitTestMethods()).ToList();
```

File: NUnit_Engine/Query/NunitTestMethods.cs [hunk 4]
```diff
@@ -20,20 +20,31 @@
...
+        }
+
+        [System.ComponentModel.Description("Returns all methods in a type that have NUnit 'Test' attribute.")]
+        [Input("nunitClass", "Type to parse in search for NUnit test methods.")]
+        [Output("methods", "All methods that have NUnit 'Test' attribute.")]
+        public static List<MethodInfo> NunitTestMethods(this Type nunitClass)
```
Comment (FraserGreenroyd) [inline NUnit_Engine/Query/NunitTestMethods.cs:45]:
```suggestion
        public static List<MethodInfo> NUnitTestMethods(this Type nunitClass)
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
Have made comments on the code for some capitalisation things on the methods I like.

I don't see the need for a `TestRun RunTests(IEnumerable<string> filePaths)` method, at least not without the original `TestRun RunTests(string filePath)` method that we had before, as that places unnecessary onerous on the calling assembly to convert a single string file path into a list just to use the method for one file path. In particular, this is what BHoMBot is using - as it executes each individual test file in sequence and never all at once.

So if we truly see a need somewhere for the `IEnumerable` option, then fine, but we also need the singular file path option to be kept as well.  Personally, I would rather do away with the IEnumerable` option and just have the singular - as it's less onerous for the calling assembly to break out their list than to force one to provide a list when they only have one (as is the case of the bot), but if there's a valid argument for an `IEnumerable` option then we can keep it providing we put the singular back in as well.

I also don't like the changing of the XML Properties around `Settings` and `Properties` - it would be good to understand better what the crash specifically was to understand how those changes fix it. I think this will break the existing set up for the bot. I'm not against updating the bot to match this framework if we have to, but I think there's a different issue possibly in play, but without any screenshots/further information on the PR/issue #435 it's hard to say. So this needs further discussion I think.

I have no qualms with the breaking out of the conversion from the XML Test Result to TestRun becoming its own method in the Convert class, that makes logical sense given the advancements we're aiming for with this PR so 👍 on that from me.

In summary, I think this is a decent start for upgrading the NUnit_Engine as we continue to grow out our testing capabilities, but a couple of things would be good to chat through next week when you're back @pawelbaran to make sure we're leveraging the right things for the right approaches (bot/BIM/other) 😄

**Comment by pawelbaran** [comment]
Thanks @FraserGreenroyd for having a look.

Regarding the addition of `RunTests` that takes a collection of strings, please note I kept the overload with a single assembly path input, but added the ability to specify classes to test. Hopefully this will not break the bot, and if it does, then happy to bring the original setup back.
image

When it comes to serialisation, the code on `develop` struggles on deserialisation of collections in my cases, so I simply made it work. I have absolutely zero opinion on the design of the code here, my only need is to pass the information from XML to test result, so happy for you to take over the steering wheel. But probs best through it together - will schedule a call in the coming days 👍

**Comment by FraserGreenroyd** [comment]
> I also don't like the changing of the XML Properties around `Settings` and `Properties`

I have reviewed this with @pawelbaran today and gone through some more tests, I now understand better why this change came in, and am happy with the change being made/kept as it does better support our workflows. Thanks @pawelbaran for being patient and working with me on this 😄

**Comment by pawelbaran** [comment]
I think we are ready for re-review @FraserGreenroyd 👍

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Comment by pawelbaran** [comment]
@FraserGreenroyd, there are some compliance checks failing, but none of the failures got introduced by this PR I think - are you fine with dispensating?

**Comment by FraserGreenroyd** [comment]
> @FraserGreenroyd, there are some compliance checks failing, but none of the failures got introduced by this PR I think - are you fine with dispensating?

Ordinarily I would agree, however, the project-compliance failures are new following our change from `C:/ProgramData` to `$(ProgramData)` and are only being picked up on a PR by PR basis - you can use the bot to fix the project file and that should resolve those suitably.

For the documentation compliance - it is that you have an input attribute that doesn't match any input parameters (`filePath` instead of the param `filePaths`) so I would suggest we resolve that one.

**Comment by pawelbaran** [comment]
> For the documentation compliance - it is that you have an input attribute that doesn't match any input parameters (`filePath` instead of the param `filePaths`) so I would suggest we resolve that one.

Yeah ofc. Not sure how did I read the test results 🤦 Thanks @FraserGreenroyd

**Comment by pawelbaran** [comment]
@BHoMBot fix project file ref. 13751724291

**Comment by pawelbaran** [comment]
@BHoMBot check compliance

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Review by FraserGreenroyd — APPROVED** [review]
Everything is in line with our discussion yesterday @pawelbaran thanks for the updates - happy to merge and deploy.
